In [1]:
import numpy as np


ejercicio 1

In [2]:
class Puzzle:
    def __init__(self, board, parent=None, move=None):
        self.board = board  # Tablero 3x3
        self.goal = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 'e']], dtype=object)  # define la meta
        self.empty_pos = tuple(np.argwhere(self.board == 'e')[0])  # Posición del espacio vacío
        self.parent = parent  # Nodo padre
        self.move = move  # Movimiento que llevó a este estado

    def is_goal(self):
        # Verifica si el estado actual es el estado objetivo
        return np.array_equal(self.board, self.goal)
        
    def print_puzzle(self):
        # Imprime el estado actual del tablero
        print(self.board)
        
    def pos_actions(self):
        # Genera los estados sucesores moviendo el espacio vacío
        successors = []
        x, y = self.empty_pos
        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]  # Arriba, abajo, izquierda, derecha
        moves = ["Up", "Down", "Left", "Right"]  # Correspondiente a las direcciones
        
        for (dx, dy), move in zip(directions, moves):
            new_x, new_y = x + dx, y + dy
            if 0 <= new_x < 3 and 0 <= new_y < 3:
                new_board = np.copy(self.board)
                new_board[x, y], new_board[new_x, new_y] = new_board[new_x, new_y], new_board[x, y]
                successors.append(Puzzle(new_board, parent=self, move=move))

        return successors
        
    def get_position(self, element):
        # Busca la posición del elemento en el tablero
        position = np.where(self.board == element)
        return position[0][0], position[1][0]
        
    def get_goal(self, element):
        # Busca la posición del elemento en el tablero meta
        position = np.where(self.goal == element)
        return position[0][0], position[1][0]
        
    def get_board(self):
        return self.board
        
    def get_heuristic(self):
        # Devuelve la heurística
        heuristic = 0
        for x in [1, 2, 3, 4, 5, 6, 7, 8, 'e']:
            heuristic += abs(self.get_position(x)[1] - self.get_goal(x)[1]) + abs(self.get_position(x)[0] - self.get_goal(x)[0])
        return heuristic
    
    def get_path(self):
        # Retrocede desde el nodo actual hasta el inicial para obtener el camino de soluciones
        path = []
        current = self
        while current.parent is not None:
            path.append(current.move)
            current = current.parent
        path.reverse()
        return path


ejercicio 2

In [12]:
#define tablero inicial
def dfs(board):
    visited = set()
    pile = []
    puzzle = Puzzle(board)
    
    # Añadir el estado inicial
    hashable_board = tuple(tuple(row) for row in board)
    visited.add(hashable_board)
    pile.append(puzzle)
    counter=0
    
    while pile:
        current_node = pile.pop() #extrae el ultimo nodo anadido
        current_board = current_node.get_board()
        current_hashable_board = tuple(tuple(row) for row in current_board)
        counter=counter+1
        # Verificar si se ha alcanzado el objetivo
        if current_node.is_goal():
            print("Objetivo alcanzado!")
            current_node.print_puzzle()
            print(counter)
            return current_node.get_path()

        
        # Expandir los nodos sucesores si no se ha alcanzado el objetivo
        for neighbor in current_node.pos_actions():
            neighbor_board = neighbor.get_board()
            neighbor_hashable_board = tuple(tuple(row) for row in neighbor_board)
            
            if neighbor_hashable_board not in visited:
                visited.add(neighbor_hashable_board)
                pile.append(neighbor)
    
    print("No se encontró solución.")
    return None
start=np.array([[1,'e',2],[6,3,4],[7,5,8]], dtype=object)
dfs(start)

Objetivo alcanzado!
[[1 2 3]
 [4 5 6]
 [7 8 'e']]
72278


['Right',
 'Down',
 'Left',
 'Left',
 'Down',
 'Right',
 'Right',
 'Up',
 'Left',
 'Left',
 'Down',
 'Right',
 'Right',
 'Up',
 'Left',
 'Left',
 'Down',
 'Right',
 'Right',
 'Up',
 'Left',
 'Left',
 'Down',
 'Right',
 'Right',
 'Up',
 'Left',
 'Left',
 'Down',
 'Right',
 'Up',
 'Right',
 'Down',
 'Left',
 'Left',
 'Up',
 'Right',
 'Right',
 'Down',
 'Left',
 'Left',
 'Up',
 'Right',
 'Right',
 'Up',
 'Left',
 'Left',
 'Down',
 'Right',
 'Right',
 'Down',
 'Left',
 'Left',
 'Up',
 'Right',
 'Right',
 'Down',
 'Left',
 'Left',
 'Up',
 'Right',
 'Right',
 'Down',
 'Left',
 'Left',
 'Up',
 'Right',
 'Right',
 'Down',
 'Left',
 'Left',
 'Up',
 'Right',
 'Right',
 'Down',
 'Left',
 'Up',
 'Right',
 'Down',
 'Left',
 'Left',
 'Up',
 'Right',
 'Right',
 'Down',
 'Left',
 'Left',
 'Up',
 'Right',
 'Right',
 'Up',
 'Left',
 'Left',
 'Down',
 'Right',
 'Right',
 'Down',
 'Left',
 'Left',
 'Up',
 'Right',
 'Right',
 'Down',
 'Left',
 'Left',
 'Up',
 'Right',
 'Right',
 'Down',
 'Left',
 'Left',
 

ejercicio 3

In [11]:
def bfs(board):
    visited = set()
    pile = []
    puzzle = Puzzle(board)
    
    # Añadir el estado inicial
    hashable_board = tuple(tuple(row) for row in board)
    visited.add(hashable_board)
    pile.append(puzzle)
    counter=0
    
    while pile:
        current_node = pile.pop(0) #extrae el primer nodo anadido
        current_board = current_node.get_board()
        current_hashable_board = tuple(tuple(row) for row in current_board)
        counter=counter+1
        
        # Verificar si se ha alcanzado el objetivo
        if current_node.is_goal():
            print("Objetivo alcanzado!")
            current_node.print_puzzle()
            print(counter)
            return current_node.get_path()
        
        # Expandir los nodos sucesores si no se ha alcanzado el objetivo
        for neighbor in current_node.pos_actions():
            neighbor_board = neighbor.get_board()
            neighbor_hashable_board = tuple(tuple(row) for row in neighbor_board)
            
            if neighbor_hashable_board not in visited:
                visited.add(neighbor_hashable_board)
                pile.append(neighbor)
    
    print("No se encontró solución.")
    return None
start=np.array([[1,'e',2],[6,3,4],[7,5,8]], dtype=object)
bfs(start)

Objetivo alcanzado!
[[1 2 3]
 [4 5 6]
 [7 8 'e']]
8351


['Right',
 'Down',
 'Left',
 'Left',
 'Up',
 'Right',
 'Right',
 'Down',
 'Left',
 'Up',
 'Left',
 'Down',
 'Right',
 'Down',
 'Right']

ejercicio 4
dfs 72,278
bfs 8,351
como podemos ver el algoritmo de primero en amplitud localiza un camino expandiendo solo una fraccion de la cantidad de nodos utilizados por el algoritm primero en profundidad

In [ ]:
def a_star(board):
    visited = set()
    pile = []
    puzzle = Puzzle(board)
    
    # Añadir el estado inicial
    hashable_board = tuple(tuple(row) for row in board)
    visited.add(hashable_board)
    pile.append(puzzle)
    counter=0
    
    while pile:
        current_node = pile.pop(0) #extrae el primer nodo anadido
        current_board = current_node.get_board()
        current_hashable_board = tuple(tuple(row) for row in current_board)
        counter=counter+1
        
        # Verificar si se ha alcanzado el objetivo
        if current_node.is_goal():
            print("Objetivo alcanzado!")
            current_node.print_puzzle()
            print(counter)
            return current_node.get_path()
        
        # Expandir los nodos sucesores si no se ha alcanzado el objetivo
        for neighbor in current_node.pos_actions():
            neighbor_board = neighbor.get_board()
            neighbor_hashable_board = tuple(tuple(row) for row in neighbor_board)
            
            if neighbor_hashable_board not in visited:
                visited.add(neighbor_hashable_board)
                pile.append(neighbor)
    
    print("No se encontró solución.")
    return None
start=np.array([[1,'e',2],[6,3,4],[7,5,8]], dtype=object)
bfs(start)